README.md
# R.cache: Fast and Light-weight Caching (Memoization) of Objects and Results to Speed Up Computations


## Installation
R package R.cache is available on [CRAN](http://cran.r-project.org/package=R.cache) and can be installed in R as:
```r
install.packages('R.cache')
```




## Software status

| Resource:     | CRAN        | Travis CI     | Appveyor         |
| ------------- | ------------------- | ------------- | ---------------- |
| _Platforms:_  | _Multiple_          | _Linux_       | _Windows_        |
| R CMD check   | <a href="http://cran.r-project.org/web/checks/check_results_R.cache.html"><img border="0" src="http://www.r-pkg.org/badges/version/R.cache" alt="CRAN version"></a> | <a href="https://travis-ci.org/HenrikBengtsson/R.cache"><img src="https://travis-ci.org/HenrikBengtsson/R.cache.svg" alt="Build status"></a> | <a href="https://ci.appveyor.com/project/HenrikBengtsson/r-cache"><img src="https://ci.appveyor.com/api/projects/status/github/HenrikBengtsson/R.cache?svg=true" alt="Build status"></a> |
| Test coverage |                     | <a href="https://coveralls.io/r/HenrikBengtsson/R.cache"><img src="https://coveralls.io/repos/HenrikBengtsson/R.cache/badge.svg?branch=develop" alt="Coverage Status"/></a>   |                  |


inst/NEWS

Package: R.cache
===============

Version: 0.5.1 [2011-08-31]
o BUG FIX (for R v2.12.2 and before): After adding support for 
  compressed files in R.cache v0.5.0, we would get the 'Error in
  seek.connection(con, origin = "current", where = -5) : whence = "end"
  is not implemented for gzfile connections' in readCacheHeader()
  iff running R v2.12.2 or before.


Version: 0.5.0 [2011-08-16]
o Added support for saving and loaded compressed cache files.
o ROBUSTNESS: Now findCache() asserts that any identified cache 
  file is really a file.
o CLEANUP: Now package uses packageStartupMessage() instead of cat().


Version: 0.4.3 [2011-07-21]
o DOCUMENTATION: Clarified how argument 'dirs' (in methods using it)
  specifies the subdirectory of the cache root directory.


Version: 0.4.2 [2011-05-19]
o Added a help page to clearCache().
o Now argument 'prompt' for clearCache() defaults to TRUE
  iff interactive() is TRUE, otherwise FALSE.


Version: 0.4.1 [2011-04-02]
o Now generateCache() utilizes a generic function getChecksum()
  to obtain the checksum.  This makes it possible to customize
  how checksums are calculated for different classes of objects.
o Added a default getChecksum() which utilizes digest::digest().
o Added getChecksum() for the Object, which first clears the
  cache of the object and the calls the default ditto.
o Added trial version of evalWithMemoization().
o BUG FIX: clearCache() would also report on subdirectories.


Version: 0.4.0 [2011-02-14]
o Added trial versions of memoizedCall() and addMemoization().


Version: 0.3.0 [2010-03-13]
o Added a NAMESPACE.


Version: 0.2.0 [2009-10-16]
o BUG FIX: In R v2.10.0 and newer, we would get an error reporting that
  internal function loadFromConn() does not exists.


Version: 0.1.9 [2009-09-11]
o Added argument 'onError' to loadCache(), to specify the action when 
  an error occurs.  The default used to be to print the error message
  (onError="print"), but now the default is to generate a warning
  ("warning").  The other alternatives are do silently ignore it, or
  to throw the error ("error").  Except for onError="error", loadCache()
  always returns NULL if an error occurs.


Version: 0.1.8 [2009-08-11]
o ROBUSTNESS: Added sanity check to readCacheHeader() testing 
  that the read header identifier is non-empty.  This results
  in a clearer error message that before.
o Now readCacheHeader() reports the "pathname" in error/warnings
  messages, if argument 'file' refers to a file and the 
  "description" if it refers to a connection.
o CLEAN UP: Renamed the HISTORY file to NEWS.


Version: 0.1.7 [2008-02-27]
o Added option to updated the "last-modified" timestamp of cache 
  files whenever they are loaded via loadCache(). This will help
  identified what cache files have not been used for a long time.
  To turn this on, use options("R.cache::touchOnLoad"=TRUE).
o Now error messages specifies the pathname, if available.
o BUG FIX: The throw() for invalid identifiers was trying to put 
  the connection object in the output and not the identifier.


Version: 0.1.6 [2007-07-02]
o Now environments, in addition to lists, may be used as keys
  for generating cache names.
o Package passes R CMD check on R v2.6.0 devel.  Minor mods.


Version: 0.1.5 [2007-03-11]
o Added getCacheRootPath() to the help index page.
o Package passes R CMD check on R v2.5.0.


Version: 0.1.4 [2007-03-07]
o Added getCacheRootPath() and setCacheRootPath() to specify
  where the root of the file path is located.


Version: 0.1.3 [2007-01-24]
o Now saveCache() and loadCache() takes optional argument 'dirs' for
  caching data to subdirectories of the root cache directory.
o Added setCacheRootPath() to set the root cache directory.


Version: 0.1.2 [2006-05-25]
o Added argument 'pathname' to loadCache() in order to load
  "unknown" cache files for which the key is unknown.
o BUG FIX: Work around for not saving "promises" (non-evaluated arguments)
  in base::save(), which otherwise includes all of the surrounding 
  environment if 'sources' is not evaluated/missing.  For more details 
  see code and my email to r-devel on 2006-05-25.  Thanks to Brian Ripley
  for explaining what was going on.


Version: 0.1.1 [2006-05-22]
o Added header comment for file format > v0.1.
o Added detection of file format version.
o Added readCacheHeader().


Version: 0.1.0 [2005-12-16]
o Using a special binary file format now.  It allows you to
  check if cache is up-to-date to a source file.  If not,
  the rest of the cache binary file is not loaded.
o Added loadCache() and saveCache() methods.
o Created.


R/loadCache.R
#########################################################################/**
# @RdocDefault loadCache
#
# @title "Loads data from file cache"
#
# \description{
#  @get "title", which is unique for an optional key object.
# }
#
# @synopsis
#
# \arguments{
#   \item{key}{An optional object from which a hexadecimal hash
#     code will be generated and appended to the filename.}
#   \item{sources}{Optional source objects.  If the cache object has a 
#     timestamp older than one of the source objects, it will be ignored 
#     and removed.}
#   \item{suffix}{A @character string to be appended to the end of the
#     filename.}
#   \item{removeOldCache}{If @TRUE and the cache is older than the
#     \code{sources}, the cache file is removed, otherwise not.}
#   \item{pathname}{The pathname to the cache file.  If specified,
#     arguments \code{key} and \code{suffix} are ignored.  Note that
#     this is only needed in order to read a cache file for which
#     the key is unknown, for instance, in order to investigate
#     an unknown cache file.}
#   \item{dirs}{A @character @vector constituting the path to the
#      cache subdirectory to be used. If @NULL, the root path is used.}
#   \item{...}{Additional argument passed to @see "base::load".}
# }
#
# \value{
#   Returns an \R object or @NULL, if cache does not exist.
# }
#
# \details{
#   The hash code calculated from the \code{key} object is a 
#   32 characters long hexadecimal MD5 hash code.  For more details,
#   see the \emph{digest} package.
# }
#
# \section{Requirements}{
#  To make use of the \code{key} argument, the \emph{digest} package 
#  (available on CRAN) must be installed, otherwise an error is generated.
#  That package is not required when \code{key==NULL}.
# }
#
# @author
#
# @examples "../incl/loadCache.Rex"
#
# \seealso{
#  @see "saveCache".
#  The @see "Biobase::cache" method in Bioconductor.
# }
#
# @keyword "programming"
# @keyword "IO"
#*/#########################################################################
setMethodS3("loadCache", "default", function(key=NULL, sources=NULL, suffix=".Rcache", removeOldCache=TRUE, pathname=NULL, dirs=NULL, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Load functions
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Same a base::load(), but it does not open a gzcon() connection, which
  # cause problem because that reset file connections to position zero.
  baseLoad <- function(con, envir=parent.frame()) {
    magic <- readChar(con, 5);
    if (regexpr("RD[AX]2\n", magic) == -1) {
      if (regexpr("RD[ABX][12]\r", magic) == 1) {
        stop("input has been corrupted, with LF replaced by CR");
      } else {
        stop("the input does not start with a magic number compatible with loading from a connection");
      }
    } else {
      .Internal(loadFromConn(con, envir));
    }
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Find cached file
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (is.null(pathname))
    pathname <- findCache(key=key, suffix=suffix, dirs=dirs);
  if (is.null(pathname))
    return(NULL);

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Try to load cached object from file connection
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (!isFile(pathname))
    return(NULL);

  fh <- file(pathname, open="rb");
  on.exit({
    if (!is.null(fh))
      close(fh);
  });

  tryCatch({
    header <- readCacheHeader(fh);
    if (!is.null(sources))
      header$sources <- sources;

    attachLocally(header);
    if (!is.null(timestamp) && !is.null(sources)) {
      for (pathname in sources) {
        if (!is.character(sources)) {
          warning("No timestamp check of cache was performed. Unsupported type of cache source: ", class(sources)[1]);
          break;
        }

        if (!file.exists(pathname)) {
          warning("No timestamp check of cache was performed. Source file not found: ", pathname);
          break;
        }

        # Is source file newer than cache?
        lastModified <- file.info(pathname)$mtime;
        if (lastModified > timestamp) {
          # Remove out-of-date cache file?
          if (removeOldCache) {
            close(fh); fh <- NULL;
            file.remove(pathname);
          }
          return(NULL);
        }
      } # for (pathname in sources)
    }

    # 4. Load cached object:
    object <- NULL;  # To please 'codetools' in R v2.6.0
    vars <- baseLoad(con=fh, ...);  # 'vars' holds names of loaded objects
    if (!identical(vars, "object")) {
      throw("Rcache file format error. Expected 'object' object: ", 
                                                paste(vars, collapse=", "));
    }

    # 5. Return cached object
    return(object);
  }, error = function(ex) {
     print(ex);
  })

  NULL;
})


############################################################################
# HISTORY:
# 2006-08-09
# o Added link to cache() in Biobase.
# 2006-05-25
# o Added argument 'pathname' to make it possible to load a cache file
#   explicitly.
# 2006-04-04
# o Added header comment for file format > v0.1.
# o Added detection of file format version.
# 2005-12-09
# o Added support for internal 'cache' and 'timestamp' objects.
# 2005-12-06
# o Created.
############################################################################


R/999.package.R
#########################################################################/**
# @RdocPackage R.cache
#
# \description{
#   @eval "getDescription(R.cache)"
#
#   \emph{Warning}: 
#   The Application Programming Interface (API) of the classes and methods
#   in this package may change.  Classes and methods are considered either 
#   to be stable, or to be in beta or alpha (pre-beta) stage.
#
#   For package history, see \code{showHistory(R.cache)}.
# }
#
# \section{Requirements}{
#   This package requires the \pkg{R.oo} [1] and \pkg{R.utils} packages.
#   For automatic generation of cache filenames, which is the most common
#   usage of this package, the CRAN package \pkg{digest} is also required.
# } 
#
# \section{Installation and updates}{
#   To install this package, see \url{http://www.braju.com/R/}.
# } 
#
# \section{To get started}{
#   \itemize{
#    \item{\link{loadCache}, \link{saveCache}}{
#      Methods for loading and saving objects from and to the cache.
#    }
#    \item{\link{getCacheRootPath}, \link{setCacheRootPath}}{
#      Methods for getting and setting the directory where cache
#      files are stored.
#    }
#   }
# }
#
# \section{How to cite this package}{
#   Whenever using this package, please cite [1] as
#
#   @howtocite "R.oo"
# }
#
# \section{Wishlist}{
#  Here is a list of features that would be useful, but which I have
#  too little time to add myself. Contributions are appreciated.
#  \itemize{
#    \item Add a functionality to identify cache files that are 
#       no longer of use.  For now, there is an extra header field
#       for arbitrary comments which can be used, but maybe more
#       formal fields are useful, e.g. keywords, user, etc?
#  }
#
#  If you consider implement some of the above, make sure it is not
#  already implemented by downloading the latest "devel" version!
# }
#
# @author
#
# \section{License}{
#   The releases of this package is licensed under 
#   LGPL version 2.1 or newer.
#
#   The development code of the packages is under a private licence 
#   (where applicable) and patches sent to the author fall under the
#   latter license, but will be, if incorporated, released under the
#   "release" license above.
# }
#
# \section{References}{
# [1] @include "../incl/BengtssonH_2003.bib.Rdoc" \cr
# }
#*/#########################################################################  


R/generateCache.R
#########################################################################/**
# @RdocDefault generateCache
#
# @title "Generates a cache pathname from a key list object"
#
# \description{
#  @get "title".
# }
#
# @synopsis
#
# \arguments{
#   \item{key}{An object from which a hexadecimal hash code will be 
#     generated and appended to the filename.}
#   \item{suffix}{A @character string to be appended to the end of the
#     filename.}
#   \item{...}{Arguments passed to @see "getCachePath".}
# }
#
# \value{
#   Returns the path as a @character string.
# }
#
# \details{
# }
#
# @author
#
# @keyword "programming"
# @keyword "IO"
# @keyword "internal"
#*/######################################################################### 
setMethodS3("generateCache", "default", function(key=NULL, suffix=".Rcache", ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'key':
  if (!is.null(key)) {
    if (!is.list(key))
      throw("Argument 'key' must be a list or NULL: ", class(key)[1]);

    # Using key object requires the CRAN package 'digest'.
    require(digest) || throw("Package not loaded: digest");
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Generate cache name from hash code of key object
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (!is.null(key)) {
    hashCode <- digest(key);
    cacheName <- hashCode;
  }

  # Add cache directory or pathname
  path <- getCachePath(...);
  cacheName <- file.path(path, cacheName);

  # Add suffix
  cacheName <- paste(cacheName, suffix, sep="");


  cacheName;
})


############################################################################
# HISTORY:
# 2005-12-09
# o Removed 'file' argument.
# 2005-12-06
# o Created.
############################################################################


R/readCacheHeader.R
#########################################################################/**
# @RdocDefault readCacheHeader
#
# @title "Loads data from file cache"
#
# \description{
#  @get "title", which is unique for an optional key object.
# }
#
# @synopsis
#
# \arguments{
#   \item{file}{A filename or a @connection.}
#   \item{...}{Not used.}
# }
#
# \value{
#   Returns a named @list structure with element \code{identifier},
#   \code{version}, \code{comment} (optional), \code{sources} (optional),
#   and \code{timestamp}.
# }
#
# @author
#
# @examples "../incl/readCacheHeader.Rex"
#
# \seealso{
#  @see "findCache".
#  @see "loadCache".
#  @see "saveCache".
# }
#
# @keyword "programming"
# @keyword "IO"
#*/######################################################################### 
setMethodS3("readCacheHeader", "default", function(file, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Load functions
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Same a base::load(), but it does not open a gzcon() connection, which
  # cause problem because that reset file connections to position zero.
  baseLoad <- function(con, envir=parent.frame()) {
    magic <- readChar(con, 5);
    if (regexpr("RD[AX]2\n", magic) == -1) {
      if (regexpr("RD[ABX][12]\r", magic) == 1) {
        stop("input has been corrupted, with LF replaced by CR");
      } else {
        stop("the input does not start with a magic number compatible with loading from a connection");
      }
    } else {
      .Internal(loadFromConn(con, envir));
    }
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'file':
  if (inherits(file, "connection")) {
  } else {
    file <- as.character(file);
    if (!isFile(file))
      throw("Argument 'file' is not an existing file: ", file);

    file <- file(file, open="rb");
    on.exit({
      if (!is.null(file))
        close(file);
    });
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Try to load cached object from file connection
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  header <- list();

  # 1a. Load identifier:
  header$identifier <- readChar(con=file, nchars=64);
  pattern <- "^Rcache v([0-9][0-9]*[.][0-9][0-9]*([.][0-9][0-9]*)*).*";
  if (regexpr(pattern, header$identifier) == -1) {
    throw("Rcache file format error. Invalid identifier: ", file);
  }

  # 1b. Get version
  header$version <- gsub(pattern, "\\1", header$identifier);

  # 1c. Read trailing '\0'.
  dummy <- readBin(con=file, what=integer(), size=1, n=1);

  # 2a. Load comment
  if (compareVersion(header$version, "0.1") > 0) {
    nchars <- readBin(con=file, what=integer(), size=4, n=1);
    header$comment <- readChar(con=file, nchars=nchars);
  }

  # 2b. Read trailing '\0'.
  dummy <- readBin(con=file, what=integer(), size=1, n=1);

  # 3. Load sources:
  sources <- NULL;  # To please 'codetools' in R v2.6.0
  vars <- baseLoad(con=file, ...);
  if (!identical(vars, "sources")) {
    throw("Rcache file format error. Expected 'sources' object: ", 
                                                paste(vars, collapse=", "));
  }
  header$sources <- sources;

  # 4. Load timestamp:
  vars <- baseLoad(con=file, ...);
  if (!identical(vars, "timestamp")) {
    throw("Rcache file format error. Expected 'timestamp' object: ", 
                                                paste(vars, collapse=", "));
  }
  header$timestamp <- timestamp;

  header;
})


############################################################################
# HISTORY:
# 2006-04-04
# o Created.
############################################################################


R/saveCache.R
#########################################################################/**
# @RdocDefault saveCache
#
# @title "Saves data to file cache"
#
# \description{
#  @get "title", which is unique for an optional key object.
# }
#
# @synopsis
#
# \arguments{
#   \item{object}{The object to be saved to file.}
#   \item{key}{An optional object from which a hexadecimal hash
#     code will be generated and appended to the filename.}
#   \item{sources}{Source objects used for comparison of timestamps when 
#     cache is loaded later.}
#   \item{suffix}{A @character string to be appended to the end of the
#     filename.}
#   \item{comment}{An optional @character string written in ASCII at the
#     beginning of the file.}
#   \item{dirs}{A @character @vector constituting the path to the
#      cache subdirectory to be used. If @NULL, the root path is used.}
#   \item{...}{Additional argument passed to @see "base::save".}
# }
#
# \value{
#   Returns (invisible) the pathname of the cache file.
# }
#
# \section{Requirements}{
#  To make use of the \code{key} argument, the \emph{digest} package 
#  (available on CRAN) must be installed, otherwise an error is generated.
#  That package is not required when \code{key==NULL}.
# }
#
# @author
#
# \examples{\dontrun{For an example, see ?loadCache}}
#
# \seealso{
#  For more details on how the hash code is generated etc, @see "loadCache".
# }
#
# @keyword "programming"
# @keyword "IO"
#*/######################################################################### 
setMethodS3("saveCache", "default", function(object, key=NULL, sources=NULL, suffix=".Rcache", comment=NULL, dirs=NULL, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Generate cache name from basename and hash object.
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  pathname <- generateCache(key=key, suffix=suffix, dirs=dirs);

  # Compression is not possible!  See loadCache().
  compress <- FALSE;

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Save to file connection
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  fh <- file(pathname, open="wb");
  on.exit(close(fh));

  # Save 'identifier' 
  identifier <- "Rcache v0.1.3 (R package R.cache by Henrik Bengtsson)";
  if (nchar(identifier) > 64)
    throw("Internal error. Identifier is too long: ", identifier);
  tail <- paste(rep(" ", 64-nchar(identifier)), collapse="");
  identifier <- paste(identifier, tail, sep="");
  writeChar(con=fh, identifier, nchars=64);

  # Save 'comment'
  if (is.null(comment))
    comment <- "";
  writeBin(con=fh, nchar(comment), size=4);
  writeChar(comment, con=fh, nchars=nchar(comment));

  # Save 'sources'
  # If 'sources' is not evaluated, it is a so called promise, which will
  # make all of its calling environments to be save too.
  dummy <- is.null(sources);
  base::save(file=fh, sources, compress=compress, ...);

  # Save 'timestamp'
  timestamp <- Sys.time();
  base::save(file=fh, timestamp, compress=compress, ...);
  
  # Save 'object'
  base::save(file=fh, object, compress=compress, ...);

  invisible(pathname);
})

############################################################################
# HISTORY:
# 2007-01-24
# o Now saveCache() returns the pathname to the cache file.
# 2006-05-25
# o BUG FIX: Work around for not saving "promises" (non-evaluated arguments)
#   in base::save(), which otherwise includes all of the surrounding 
#   environment if 'sources' is not evaluated/missing.  For more details 
#   see code and my email to r-devel on 2006-05-25.  Thanks to Brian Ripley
#   for explaining what was going on.
# 2006-04-04
# o Added header comment.
# 2005-12-09
# o Object save to file is now a structure containing the object to be
#   cached, a timestamp specifying the it was cached, and a source object.
# o Replaced argument 'file' with 'source'.
# 2005-12-06
# o Created.
############################################################################
